lib/matrix.rb

```diff
@@ -303,12 +303,139 @@ def [](i, j)
   alias element []
   alias component []
 
+  #
+  # :call-seq:
+  #   matrix[range, range] = new_matrix
+  #   matrix[range, range] = new_element
+  #   matrix[range, integer] = vector
+  #   matrix[range, integer] = column_matrix
+  #   matrix[range, integer] = new_element
+  #   matrix[integer, range] = vector
+  #   matrix[integer, range] = row_matrix
+  #   matrix[integer, range] = new_element
+  #   matrix[integer, integer] = new_element
+  #
+  # Set element or elements of matrix.
   def []=(i, j, v)
-    @rows[i][j] = v
+    raise FrozenError, "can't modify frozen Matrix" if frozen?
+    if both_ranges?(i, j)
+      i = convert_range(i, row_count) if i.begin > i.end
+      j = convert_range(j, column_count) if j.begin > j.end
+      raise IndexError, "expected ranges are outside of matrix" unless range_within_matrix_range?(i, j)
+      set_row_and_col_range(i, j, v)
+    elsif one_is_range?(i, j)
+      set_range(i, j, v)
+    else
+      i = CoercionHelper.coerce_to_int(i)
+      j = CoercionHelper.coerce_to_int(j)
+      raise IndexError, "indices are outside of matrix" unless indices_within_matrix?(i, j)
+      set_value(i, j, v)
+    end
   end
   alias set_element []=
   alias set_component []=
-  private :[]=, :set_element, :set_component
```

Comment:
`set_element` and `set_component` should remain private. They are there only for historical reasons.